File: fontend/lib/student.ts
/**
 * Student console API client — the sole frontend data boundary for
 * C-ST-01 … C-ST-20.
 *
 * No function here takes a student id. §4.9 scopes a learner to their own
 * data, and the API enforces that by resolving the caller's enrolment
 * server-side — so there is no parameter for the browser to tamper with.
 */

import { API_BASE_URL, getAccessToken } from "./auth";
import { APIError, requestJson } from "./api-client";
import { queryString } from "./principal";

const API_PREFIX = "student";

export { APIError as StudentAPIError };
export { queryString };

const call = <T>(path: string, init: RequestInit = {}): Promise<T> =>
  requestJson<T>(
    `${API_BASE_URL}/api/v1/${API_PREFIX}${path}`,
    init,
    getAccessToken(),
    "StudentAPIError",
  );

const body = (payload: unknown) => JSON.stringify(payload);

// ── Shared shapes ───────────────────────────────────────────────────────────

export interface StudentPage<T> {
  total: number;
  limit: number;
  offset: number;
  items: T[];
}

export interface StudentSubjectOption {
  id: string;
  code: string;
  name: string;
  subject_type: string;
  teacher_names: string[];
}

export interface StudentNoticeRow {
  id: string;
  title: string;
  body: string;
  author_name: string | null;
  target_scope: string;
  priority: "NORMAL" | "IMPORTANT" | "URGENT";
  is_pinned: boolean;
  published_at: string;
  expires_at: string | null;
  is_read: boolean;
}

// ── C-ST-01 dashboard ───────────────────────────────────────────────────────

export interface StudentTodayClass {
  slot_id: string;
  subject_id: string | null;
  subject_code: string | null;
  subject_name: string | null;
  teacher_name: string | null;
  period_number: number;
  start_time: string;
  end_time: string;
  room_no: string | null;
  slot_type: string;
}

export interface StudentUpcomingExam {
  id: string;
  title: string;
  subject_code: string;
  subject_name: string;
  exam_type: string;
  mode: string;
  scheduled_at: string;
  window_end_at: string | null;
  duration_minutes: number;
  total_marks: number;
  status: string;
  attempt_status: string | null;
  can_attempt: boolean;
}

export interface StudentPendingAssignment {
  id: string;
  title: string;
  subject_code: string;
  due_date: string;
  is_overdue: boolean;
  status: string;
}

export interface StudentDashboard {
  academic_year: string | null;
  class_name: string;
  roll_number: string | null;
  today: string;
  attendance_percentage: number | null;
  attendance_threshold: number | null;
  is_attendance_short: boolean;
  today_classes: StudentTodayClass[];
  pending_assignment_count: number;
  pending_assignments: StudentPendingAssignment[];
  upcoming_exam_count: number;
  upcoming_exams: StudentUpcomingExam[];
  unread_notice_count: number;
  recent_notices: StudentNoticeRow[];
  fee_balance_due: number | null;
  subjects: StudentSubjectOption[];
}

// ── C-ST-02 profile ─────────────────────────────────────────────────────────

export interface StudentProfile {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  avatar_url: string | null;
  gender: string | null;
  date_of_birth: string | null;
  address: string | null;
  roll_number: string | null;
  class_id: string;
  class_name: string;
  department_name: string | null;
  academic_year: string | null;
  enrollment_date: string | null;
  enrollment_status: string;
  mentor_name: string | null;
}

// ── C-ST-03 / C-ST-04 attendance ────────────────────────────────────────────

export interface StudentSubjectAttendance {
  subject_id: string;
  subject_code: string;
  subject_name: string;
  present_count: number;
  absent_count: number;
  late_count: number;
  excused_count: number;
  total_sessions: number;
  attendance_percentage: number | null;
  is_short: boolean;
}

export interface StudentAttendanceDay {
  date: string;
  status: "PRESENT" | "ABSENT";
  present_count: number;
  absent_count: number;
}

export interface StudentAttendanceOverview {
  from_date: string;
  to_date: string;
  attendance_percentage: number | null;
  attendance_threshold: number | null;
  is_short: boolean;
  total_sessions: number;
  present_count: number;
  absent_count: number;
  subjects: StudentSubjectAttendance[];
  days: StudentAttendanceDay[];
}

export interface StudentLeaveRow {
  id: string;
  from_date: string;
  to_date: string;
  total_days: number;
  reason: string;
  document_url: string | null;
  status: "PENDING" | "APPROVED" | "REJECTED" | "CANCELLED";
  reviewed_by_name: string | null;
  reviewed_at: string | null;
  created_at: string;
}

// ── C-ST-06 timetable ───────────────────────────────────────────────────────

export interface StudentTimetableSlot {
  id: string;
  day_of_week: number;
  period_number: number;
  start_time: string;
  end_time: string;
  subject_id: string | null;
  subject_code: string | null;
  subject_name: string | null;
  teacher_name: string | null;
  room_no: string | null;
  slot_type: string;
}

export interface StudentTimetable {
  class_name: string;
  academic_year: string | null;
  slots: StudentTimetableSlot[];
}

// ── C-ST-08 attempt screen ──────────────────────────────────────────────────

export interface StudentAttemptQuestionOption {
  id: string;
  text: string;
  sort_order: number;
}

export interface StudentAttemptQuestion {
  id: string;
  text: string;
  question_type: string;
  marks: number;
  negative_marks: number;
  image_url: string | null;
  sort_order: number;
  options: StudentAttemptQuestionOption[];
  selected_option_id: string | null;
  text_answer: string | null;
}

export interface StudentAttemptScreen {
  attempt_id: string;
  exam_id: string;
  title: string;
  subject_code: string;
  instructions: string | null;
  total_marks: number;
  duration_minutes: number;
  started_at: string;
  /** Server-computed deadline; the browser clock is advisory only. */
  expires_at: string;
  server_time: string;
  tab_switch_count: number;
  is_submitted: boolean;
  questions: StudentAttemptQuestion[];
}

export interface StudentAnswerInput {
  question_id: string;
  selected_option_id?: string | null;
  text_answer?: string | null;
}

// ── C-ST-09 exam result ─────────────────────────────────────────────────────

export interface StudentExamResultAnswer {
  question_id: string;
  question_text: string;
  question_type: string;
  question_marks: number;
  your_answer: string | null;
  correct_answer: string | null;
  score: number | null;
  feedback: string | null;
  explanation: string | null;
}

export interface StudentExamResult {
  exam_id: string;
  title: string;
  subject_code: string;
  subject_name: string;
  total_marks: number;
  passing_marks: number;
  submitted_at: string | null;
  total_score: number | null;
  percentage: number | null;
  grade: string | null;
  is_pass: boolean | null;
  status: string;
  review_available: boolean;
  answers: StudentExamResultAnswer[];
}

// ── C-ST-10 … C-ST-12 assignments ───────────────────────────────────────────

export interface StudentMilestoneRow {
  id: string;
  title: string;
  description: string | null;
  marks: number;
  due_date: string | null;
  sort_order: number;
  is_locked: boolean;
  submission_id: string | null;
  submission_status: string | null;
  score: number | null;
  feedback: string | null;
}

export interface StudentSubmissionFile {
  id: string;
  file_name: string;
  file_key: string;
  file_size_bytes: number;
  mime_type: string;
  uploaded_at: string;
}

export interface StudentSubmissionRow {
  id: string;
  milestone_id: string | null;
  milestone_title: string | null;
  text_response: string | null;
  submitted_at: string;
  is_late: boolean;
  late_by_minutes: number | null;
  score: number | null;
  grade: string | null;
  feedback: string | null;
  status: string;
  version: number;
  reviewed_at: string | null;
  files: StudentSubmissionFile[];
}

export interface StudentAssignmentRow {
  id: string;
  title: string;
  subject_id: string;
  subject_code: string;
  subject_name: string;
  teacher_name: string | null;
  assignment_type: string;
  total_marks: number;
  passing_marks: number;
  due_date: string;
  is_overdue: boolean;
  allow_late_submission: boolean;
  late_penalty_percent: number;
  my_status: string;
  my_score: number | null;
  can_submit: boolean;
}

export interface StudentAssignmentPage extends StudentPage<StudentAssignmentRow> {
  pending_count: number;
  submitted_count: number;
}

export interface StudentAssignmentDetail extends StudentAssignmentRow {
  description: string;
  max_file_size_mb: number;
  allowed_file_types: string[];
  instructions_url: string | null;
  milestones: StudentMilestoneRow[];
  submissions: StudentSubmissionRow[];
}

// ── C-ST-13 / C-ST-14 content ───────────────────────────────────────────────

export interface StudentContentRow {
  id: string;
  title: string;
  description: string | null;
  subject_id: string;
  subject_code: string;
  subject_name: string;
  content_type: string;
  file_key: string | null;
  external_url: string | null;
  file_size_bytes: number | null;
  duration_seconds: number | null;
  chapter: string | null;
  uploaded_by_name: string | null;
  created_at: string;
}

export interface StudentContentPage extends StudentPage<StudentContentRow> {
  chapters: string[];
  subjects: StudentSubjectOption[];
}

// ── C-ST-15 … C-ST-17 results ───────────────────────────────────────────────

export interface StudentResultRow {
  id: string;
  publication_id: string;
  publication_title: string;
  published_at: string;
  total_marks_obtained: number;
  total_marks_possible: number;
  percentage: number;
  grade: string;
  rank: number | null;
  result: string;
}

export interface StudentResultSubject {
  subject_code: string | null;
  subject_name: string | null;
  marks_obtained: number | null;
  marks_possible: number | null;
  grade: string | null;
}

export interface StudentResultDetail extends StudentResultRow {
  class_name: string;
  remarks: string | null;
  subjects: StudentResultSubject[];
}

// ── C-ST-19 discussion ──────────────────────────────────────────────────────

export interface StudentThreadRow {
  id: string;
  title: string;
  body: string;
  author_id: string;
  author_name: string | null;
  scope_type: "CLASS" | "SUBJECT";
  scope_id: string;
  scope_name: string | null;
  tags: string[];
  is_pinned: boolean;
  is_locked: boolean;
  is_resolved: boolean;
  reply_count: number;
  upvote_count: number;
  has_upvoted: boolean;
  is_mine: boolean;
  created_at: string;
  updated_at: string;
}

export interface StudentReplyRow {
  id: string;
  author_id: string;
  author_name: string | null;
  body: string;
  is_accepted_answer: boolean;
  upvote_count: number;
  has_upvoted: boolean;
  is_mine: boolean;
  created_at: string;
}

export interface StudentThreadDetail extends StudentThreadRow {
  replies: StudentReplyRow[];
}

// ── C-ST-20 fees ────────────────────────────────────────────────────────────

export interface StudentInstallmentRow {
  id: string;
  installment_number: number;
  label: string;
  amount: number;
  due_date: string;
  paid_amount: number;
  late_fine: number;
  status: string;
  is_overdue: boolean;
}

export interface StudentPaymentRow {
  id: string;
  amount: number;
  payment_mode: string;
  transaction_reference: string | null;
  payment_date: string;
  receipt_number: string;
  notes: string | null;
}

export interface StudentFeeAccountView {
  has_account: boolean;
  academic_year: string | null;
  total_fee: number | null;
  concession_amount: number | null;
  scholarship_amount: number | null;
  net_payable: number | null;
  total_paid: number | null;
  balance_due: number | null;
  status: string | null;
  installments: StudentInstallmentRow[];
  payments: StudentPaymentRow[];
}

// ── Endpoints ───────────────────────────────────────────────────────────────

export const fetchStudentDashboard = () => call<StudentDashboard>("/dashboard");
export const fetchStudentProfile = () => call<StudentProfile>("/profile");
export const updateStudentProfile = (payload: {
  phone?: string | null;
  address?: string | null;
  avatar_url?: string | null;
}) => call<StudentProfile>("/profile", { method: "PATCH", body: body(payload) });

export const fetchStudentAttendance = (filters: { fromDate?: string; toDate?: string } = {}) =>
  call<StudentAttendanceOverview>(
    `/attendance${queryString({ from_date: filters.fromDate, to_date: filters.toDate })}`,
  );

export const fetchStudentLeaves = (filters: { limit?: number; offset?: number } = {}) =>
  call<StudentPage<StudentLeaveRow>>(`/attendance/leaves${queryString(filters)}`);

export const applyStudentLeave = (payload: {
  from_date: string;
  to_date: string;
  reason: string;
  document_url?: string | null;
}) => call<StudentLeaveRow>("/attendance/leaves", { method: "POST", body: body(payload) });

export const cancelStudentLeave = (id: string) =>
  call<StudentLeaveRow>(`/attendance/leaves/${id}`, { method: "DELETE" });

export const fetchStudentTimetable = () => call<StudentTimetable>("/timetable");

export const fetchStudentExams = (filters: {
  status?: string;
  limit?: number;
  offset?: number;
} = {}) => call<StudentPage<StudentUpcomingExam>>(`/examinations${queryString(filters)}`);

export const startStudentAttempt = (examId: string) =>
  call<StudentAttemptScreen>(`/examinations/${examId}/attempt`, { method: "POST" });

export const saveStudentAnswers = (attemptId: string, answers: StudentAnswerInput[]) =>
  call<StudentAttemptScreen>(`/attempts/${attemptId}`, {
    method: "PATCH",
    body: body({ answers }),
  });

export const reportStudentTabSwitch = (attemptId: string, count = 1) =>
  call<StudentAttemptScreen>(`/attempts/${attemptId}/tab-switch`, {
    method: "POST",
    body: body({ count }),
  });

export const submitStudentAttempt = (attemptId: string, answers: StudentAnswerInput[] = []) =>
  call<StudentExamResult>(`/attempts/${attemptId}/submit`, {
    method: "POST",
    body: body({ answers }),
  });

export const fetchStudentExamResult = (examId: string) =>
  call<StudentExamResult>(`/examinations/${examId}/result`);

export const fetchStudentAssignments = (filters: {
  status?: string;
  limit?: number;
  offset?: number;
} = {}) => call<StudentAssignmentPage>(`/assignments${queryString(filters)}`);

export const fetchStudentAssignment = (id: string) =>
  call<StudentAssignmentDetail>(`/assignments/${id}`);

export const submitStudentAssignment = (
  id: string,
  payload: {
    milestone_id?: string | null;
    text_response?: string | null;
    files?: {
      file_name: string;
      file_key: string;
      file_size_bytes: number;
      mime_type: string;
    }[];
  },
) =>
  call<StudentAssignmentDetail>(`/assignments/${id}/submissions`, {
    method: "POST",
    body: body(payload),
  });

export const fetchStudentContent = (filters: {
  subjectId?: string;
  chapter?: string;
  contentType?: string;
  limit?: number;
  offset?: number;
} = {}) =>
  call<StudentContentPage>(
    `/content${queryString({
      subject_id: filters.subjectId,
      chapter: filters.chapter,
      content_type: filters.contentType,
      limit: filters.limit,
      offset: filters.offset,
    })}`,
  );

export const openStudentContent = (id: string) => call<StudentContentRow>(`/content/${id}`);

export const fetchStudentResults = () => call<{ items: StudentResultRow[] }>("/results");
export const fetchStudentResult = (id: string) => call<StudentResultDetail>(`/results/${id}`);

export const fetchStudentNotices = (filters: { limit?: number; offset?: number } = {}) =>
  call<StudentPage<StudentNoticeRow> & { unread_count: number }>(
    `/notices${queryString(filters)}`,
  );

export const markStudentNoticeRead = (id: string) =>
  call<Record<string, never>>(`/notices/${id}/read`, { method: "POST" });

export const fetchStudentThreads = (filters: {
  query?: string;
  limit?: number;
  offset?: number;
} = {}) => call<StudentPage<StudentThreadRow>>(`/discussion${queryString(filters)}`);

export const fetchStudentThread = (id: string) => call<StudentThreadDetail>(`/discussion/${id}`);

export const createStudentThread = (payload: {
  title: string;
  body: string;
  subject_id?: string | null;
  tags?: string[];
}) => call<StudentThreadDetail>("/discussion", { method: "POST", body: body(payload) });

export const replyToStudentThread = (id: string, text: string) =>
  call<StudentThreadDetail>(`/discussion/${id}/replies`, {
    method: "POST",
    body: body({ body: text }),
  });

export const voteOnStudentDiscussion = (
  targetType: "THREAD" | "REPLY",
  targetId: string,
) =>
  call<StudentThreadDetail>("/discussion/vote", {
    method: "POST",
    body: body({ target_type: targetType, target_id: targetId }),
  });

export const fetchStudentFees = () => call<StudentFeeAccountView>("/fees");

File: backend/app/schemas/student.py
"""Wire contracts for the Student console (C-ST-01 … C-ST-20).

§4.9 limits a student to **their own data**, so nothing here accepts a student
id: the caller *is* the subject of every request. Ids that do appear (an exam,
an assignment, a result) are checked against the caller's enrolment before a
row is returned.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime, time
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from app.schemas.common import APIResponse


class StudentPage(BaseModel):
    total: int
    limit: int
    offset: int


class StudentSubjectOption(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    subject_type: str
    teacher_names: list[str] = Field(default_factory=list)


# ── C-ST-01 dashboard / C-ST-02 profile ──────────────────────────────────────


class StudentTodayClass(BaseModel):
    slot_id: uuid.UUID
    subject_id: uuid.UUID | None = None
    subject_code: str | None = None
    subject_name: str | None = None
    teacher_name: str | None = None
    period_number: int
    start_time: time
    end_time: time
    room_no: str | None = None
    slot_type: str


class StudentNoticeRow(BaseModel):
    id: uuid.UUID
    title: str
    body: str
    author_name: str | None = None
    target_scope: str
    priority: str
    is_pinned: bool
    published_at: datetime
    expires_at: datetime | None = None
    is_read: bool = False


class StudentUpcomingExam(BaseModel):
    id: uuid.UUID
    title: str
    subject_code: str
    subject_name: str
    exam_type: str
    mode: str
    scheduled_at: datetime
    window_end_at: datetime | None = None
    duration_minutes: int
    total_marks: int
    status: str
    attempt_status: str | None = None
    can_attempt: bool = False


class StudentPendingAssignment(BaseModel):
    id: uuid.UUID
    title: str
    subject_code: str
    due_date: datetime
    is_overdue: bool
    status: str


class StudentDashboard(BaseModel):
    academic_year: str | None = None
    class_name: str
    roll_number: str | None = None
    today: date
    attendance_percentage: float | None = None
    attendance_threshold: int | None = None
    is_attendance_short: bool = False
    today_classes: list[StudentTodayClass] = Field(default_factory=list)
    pending_assignment_count: int = 0
    pending_assignments: list[StudentPendingAssignment] = Field(default_factory=list)
    upcoming_exam_count: int = 0
    upcoming_exams: list[StudentUpcomingExam] = Field(default_factory=list)
    unread_notice_count: int = 0
    recent_notices: list[StudentNoticeRow] = Field(default_factory=list)
    fee_balance_due: float | None = None
    subjects: list[StudentSubjectOption] = Field(default_factory=list)


class StudentProfile(BaseModel):
    id: uuid.UUID
    name: str
    email: str | None = None
    phone: str | None = None
    avatar_url: str | None = None
    gender: str | None = None
    date_of_birth: date | None = None
    address: str | None = None
    roll_number: str | None = None
    class_id: uuid.UUID
    class_name: str
    department_name: str | None = None
    academic_year: str | None = None
    enrollment_date: date | None = None
    enrollment_status: str
    mentor_name: str | None = None


class StudentProfileUpdate(BaseModel):
    """The fields §4.9 lets a learner maintain themselves.

    Name, roll number and class are institution records — changing them is an
    admin action, so they are deliberately absent.
    """

    phone: str | None = Field(default=None, max_length=20)
    address: str | None = Field(default=None, max_length=1000)
    avatar_url: str | None = Field(default=None, max_length=2000)


# ── C-ST-03 / C-ST-04 attendance ─────────────────────────────────────────────


class StudentSubjectAttendance(BaseModel):
    subject_id: uuid.UUID
    subject_code: str
    subject_name: str
    present_count: int
    absent_count: int
    late_count: int
    excused_count: int
    total_sessions: int
    attendance_percentage: float | None = None
    is_short: bool = False


class StudentAttendanceDay(BaseModel):
    date: date
    #: Worst status of the day — one absence matters more than three presents.
    status: str
    present_count: int
    absent_count: int


class StudentAttendanceOverview(BaseModel):
    from_date: date
    to_date: date
    attendance_percentage: float | None = None
    attendance_threshold: int | None = None
    is_short: bool = False
    total_sessions: int
    present_count: int
    absent_count: int
    subjects: list[StudentSubjectAttendance] = Field(default_factory=list)
    days: list[StudentAttendanceDay] = Field(default_factory=list)


class StudentLeaveRow(BaseModel):
    id: uuid.UUID
    from_date: date
    to_date: date
    total_days: int
    reason: str
    document_url: str | None = None
    status: str
    reviewed_by_name: str | None = None
    reviewed_at: datetime | None = None
    created_at: datetime


class StudentLeavePage(StudentPage):
    items: list[StudentLeaveRow] = Field(default_factory=list)


class StudentLeaveCreate(BaseModel):
    from_date: date
    to_date: date
    reason: str = Field(min_length=5, max_length=2000)
    document_url: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def _validate(self) -> "StudentLeaveCreate":
        if self.from_date > self.to_date:
            raise ValueError("from_date must be on or before to_date")
        if (self.to_date - self.from_date).days > 90:
            raise ValueError("A single leave request cannot span more than 90 days")
        return self


# ── C-ST-06 timetable ────────────────────────────────────────────────────────


class StudentTimetableSlot(BaseModel):
    id: uuid.UUID
    day_of_week: int
    period_number: int
    start_time: time
    end_time: time
    subject_id: uuid.UUID | None = None
    subject_code: str | None = None
    subject_name: str | None = None
    teacher_name: str | None = None
    room_no: str | None = None
    slot_type: str


class StudentTimetable(BaseModel):
    class_name: str
    academic_year: str | None = None
    slots: list[StudentTimetableSlot] = Field(default_factory=list)


# ── C-ST-07 / C-ST-08 / C-ST-09 examinations ─────────────────────────────────


class StudentExamPage(StudentPage):
    items: list[StudentUpcomingExam] = Field(default_factory=list)


class StudentAttemptQuestionOption(BaseModel):
    id: uuid.UUID
    text: str
    sort_order: int


class StudentAttemptQuestion(BaseModel):
    id: uuid.UUID
    text: str
    question_type: str
    marks: float
    negative_marks: float
    image_url: str | None = None
    sort_order: int
    options: list[StudentAttemptQuestionOption] = Field(default_factory=list)
    selected_option_id: uuid.UUID | None = None
    text_answer: str | None = None


class StudentAttemptScreen(BaseModel):
    attempt_id: uuid.UUID
    exam_id: uuid.UUID
    title: str
    subject_code: str
    instructions: str | None = None
    total_marks: int
    duration_minutes: int
    started_at: datetime
    #: Server-computed deadline. The browser clock is advisory only — the
    #: submit path re-checks this before accepting answers.
    expires_at: datetime
    server_time: datetime
    tab_switch_count: int
    is_submitted: bool
    questions: list[StudentAttemptQuestion] = Field(default_factory=list)


class StudentAnswerInput(BaseModel):
    question_id: uuid.UUID
    selected_option_id: uuid.UUID | None = None
    text_answer: str | None = Field(default=None, max_length=20_000)


class StudentAnswerSave(BaseModel):
    answers: list[StudentAnswerInput] = Field(min_length=1, max_length=200)


class StudentAttemptSubmit(BaseModel):
    answers: list[StudentAnswerInput] = Field(default_factory=list)


class StudentTabSwitch(BaseModel):
    """Reported by the attempt screen when the tab loses focus."""

    count: int = Field(default=1, ge=1, le=10)


class StudentExamResultAnswer(BaseModel):
    question_id: uuid.UUID
    question_text: str
    question_type: str
    question_marks: float
    your_answer: str | None = None
    correct_answer: str | None = None
    score: float | None = None
    feedback: str | None = None
    explanation: str | None = None


class StudentExamResult(BaseModel):
    exam_id: uuid.UUID
    title: str
    subject_code: str
    subject_name: str
    total_marks: int
    passing_marks: int
    submitted_at: datetime | None = None
    total_score: float | None = None
    percentage: float | None = None
    grade: str | None = None
    is_pass: bool | None = None
    status: str
    #: False until grading completes or the teacher enables review.
    review_available: bool = False
    answers: list[StudentExamResultAnswer] = Field(default_factory=list)


# ── C-ST-10 … C-ST-12 assignments ────────────────────────────────────────────


class StudentMilestoneRow(BaseModel):
    id: uuid.UUID
    title: str
    description: str | None = None
    marks: int
    due_date: datetime | None = None
    sort_order: int
    #: A stage stays locked until the previous one is approved (C-ST-12).
    is_locked: bool
    submission_id: uuid.UUID | None = None
    submission_status: str | None = None
    score: float | None = None
    feedback: str | None = None


class StudentSubmissionFile(BaseModel):
    id: uuid.UUID
    file_name: str
    file_key: str
    file_size_bytes: int
    mime_type: str
    uploaded_at: datetime


class StudentSubmissionRow(BaseModel):
    id: uuid.UUID
    milestone_id: uuid.UUID | None = None
    milestone_title: str | None = None
    text_response: str | None = None
    submitted_at: datetime
    is_late: bool
    late_by_minutes: int | None = None
    score: float | None = None
    grade: str | None = None
    feedback: str | None = None
    status: str
    version: int
    reviewed_at: datetime | None = None
    files: list[StudentSubmissionFile] = Field(default_factory=list)


class StudentAssignmentRow(BaseModel):
    id: uuid.UUID
    title: str
    subject_id: uuid.UUID
    subject_code: str
    subject_name: str
    teacher_name: str | None = None
    assignment_type: str
    total_marks: int
    passing_marks: int
    due_date: datetime
    is_overdue: bool
    allow_late_submission: bool
    late_penalty_percent: int
    #: PENDING | SUBMITTED | UNDER_REVIEW | APPROVED | REJECTED | RESUBMIT_REQUESTED
    my_status: str
    my_score: float | None = None
    can_submit: bool


class StudentAssignmentPage(StudentPage):
    pending_count: int = 0
    submitted_count: int = 0
    items: list[StudentAssignmentRow] = Field(default_factory=list)


class StudentAssignmentDetail(StudentAssignmentRow):
    description: str
    max_file_size_mb: int
    allowed_file_types: list[str] = Field(default_factory=list)
    instructions_url: str | None = None
    milestones: list[StudentMilestoneRow] = Field(default_factory=list)
    submissions: list[StudentSubmissionRow] = Field(default_factory=list)


class StudentSubmissionFileInput(BaseModel):
    file_name: str = Field(min_length=1, max_length=255)
    file_key: str = Field(min_length=1, max_length=2000)
    file_size_bytes: int = Field(ge=0)
    mime_type: str = Field(min_length=1, max_length=100)


class StudentSubmissionCreate(BaseModel):
    milestone_id: uuid.UUID | None = None
    text_response: str | None = Field(default=None, max_length=20_000)
    files: list[StudentSubmissionFileInput] = Field(default_factory=list, max_length=10)

    @model_validator(mode="after")
    def _validate(self) -> "StudentSubmissionCreate":
        if not (self.text_response or "").strip() and not self.files:
            raise ValueError("Attach a file or write a response before submitting")
        return self


# ── C-ST-13 / C-ST-14 content ────────────────────────────────────────────────


class StudentContentRow(BaseModel):
    id: uuid.UUID
    title: str
    description: str | None = None
    subject_id: uuid.UUID
    subject_code: str
    subject_name: str
    content_type: str
    file_key: str | None = None
    external_url: str | None = None
    file_size_bytes: int | None = None
    duration_seconds: int | None = None
    chapter: str | None = None
    uploaded_by_name: str | None = None
    created_at: datetime


class StudentContentPage(StudentPage):
    chapters: list[str] = Field(default_factory=list)
    subjects: list[StudentSubjectOption] = Field(default_factory=list)
    items: list[StudentContentRow] = Field(default_factory=list)


# ── C-ST-15 … C-ST-17 results ────────────────────────────────────────────────


class StudentResultRow(BaseModel):
    id: uuid.UUID
    publication_id: uuid.UUID
    publication_title: str
    published_at: datetime
    total_marks_obtained: float
    total_marks_possible: float
    percentage: float
    grade: str
    rank: int | None = None
    result: str


class StudentResultSubject(BaseModel):
    subject_code: str | None = None
    subject_name: str | None = None
    marks_obtained: float | None = None
    marks_possible: float | None = None
    grade: str | None = None


class StudentResultDetail(StudentResultRow):
    class_name: str
    remarks: str | None = None
    subjects: list[StudentResultSubject] = Field(default_factory=list)


class StudentResultList(BaseModel):
    items: list[StudentResultRow] = Field(default_factory=list)


# ── C-ST-18 notices ──────────────────────────────────────────────────────────


class StudentNoticePage(StudentPage):
    unread_count: int = 0
    items: list[StudentNoticeRow] = Field(default_factory=list)


# ── C-ST-19 discussion ───────────────────────────────────────────────────────


class StudentThreadRow(BaseModel):
    id: uuid.UUID
    title: str
    body: str
    author_id: uuid.UUID
    author_name: str | None = None
    scope_type: str
    scope_id: uuid.UUID
    scope_name: str | None = None
    tags: list[str] = Field(default_factory=list)
    is_pinned: bool
    is_locked: bool
    is_resolved: bool
    reply_count: int
    upvote_count: int
    has_upvoted: bool = False
    is_mine: bool = False
    created_at: datetime
    updated_at: datetime


class StudentThreadPage(StudentPage):
    items: list[StudentThreadRow] = Field(default_factory=list)


class StudentReplyRow(BaseModel):
    id: uuid.UUID
    author_id: uuid.UUID
    author_name: str | None = None
    body: str
    is_accepted_answer: bool
    upvote_count: int
    has_upvoted: bool = False
    is_mine: bool = False
    created_at: datetime


class StudentThreadDetail(StudentThreadRow):
    replies: list[StudentReplyRow] = Field(default_factory=list)


class StudentThreadCreate(BaseModel):
    title: str = Field(min_length=3, max_length=255)
    body: str = Field(min_length=1, max_length=20_000)
    subject_id: uuid.UUID | None = None
    tags: list[str] = Field(default_factory=list, max_length=8)


class StudentReplyCreate(BaseModel):
    body: str = Field(min_length=1, max_length=20_000)


class StudentVote(BaseModel):
    target_type: Literal["THREAD", "REPLY"]
    target_id: uuid.UUID


# ── C-ST-20 fees ─────────────────────────────────────────────────────────────


class StudentInstallmentRow(BaseModel):
    id: uuid.UUID
    installment_number: int
    label: str
    amount: float
    due_date: date
    paid_amount: float
    late_fine: float
    status: str
    is_overdue: bool


class StudentPaymentRow(BaseModel):
    id: uuid.UUID
    amount: float
    payment_mode: str
    transaction_reference: str | None = None
    payment_date: date
    receipt_number: str
    notes: str | None = None


class StudentFeeAccountView(BaseModel):
    has_account: bool
    academic_year: str | None = None
    total_fee: float | None = None
    concession_amount: float | None = None
    scholarship_amount: float | None = None
    net_payable: float | None = None
    total_paid: float | None = None
    balance_due: float | None = None
    status: str | None = None
    installments: list[StudentInstallmentRow] = Field(default_factory=list)
    payments: list[StudentPaymentRow] = Field(default_factory=list)


# ── API envelope aliases ─────────────────────────────────────────────────────

APIResponseStudentDashboard = APIResponse[StudentDashboard]
APIResponseStudentProfile = APIResponse[StudentProfile]
APIResponseStudentAttendance = APIResponse[StudentAttendanceOverview]
APIResponseStudentLeavePage = APIResponse[StudentLeavePage]
APIResponseStudentLeaveRow = APIResponse[StudentLeaveRow]
APIResponseStudentTimetable = APIResponse[StudentTimetable]
APIResponseStudentExamPage = APIResponse[StudentExamPage]
APIResponseStudentAttemptScreen = APIResponse[StudentAttemptScreen]
APIResponseStudentExamResult = APIResponse[StudentExamResult]
APIResponseStudentAssignmentPage = APIResponse[StudentAssignmentPage]
APIResponseStudentAssignmentDetail = APIResponse[StudentAssignmentDetail]
APIResponseStudentContentPage = APIResponse[StudentContentPage]
APIResponseStudentContentRow = APIResponse[StudentContentRow]
APIResponseStudentResultList = APIResponse[StudentResultList]
APIResponseStudentResultDetail = APIResponse[StudentResultDetail]
APIResponseStudentNoticePage = APIResponse[StudentNoticePage]
APIResponseStudentThreadPage = APIResponse[StudentThreadPage]
APIResponseStudentThreadDetail = APIResponse[StudentThreadDetail]
APIResponseStudentFees = APIResponse[StudentFeeAccountView]
APIResponseStudentEmpty = APIResponse[dict]
